File: fix_cpp_code.py
#!/usr/bin/env python3
"""
修复Fluxus生成的C++代码中的问题
"""

import re
import sys
import os

def fix_cpp_code(cpp_file_path):
    """修复C++代码中的fmt.Println问题"""

    if not os.path.exists(cpp_file_path):
        print(f"错误：文件 {cpp_file_path} 不存在")
        return False

    # 读取原始文件
    with open(cpp_file_path, 'r') as f:
        content = f.read()

    original_content = content

    print("=== 开始修复C++代码 ===")
    print(f"原始文件: {cpp_file_path}")
    print(f"原始文件大小: {len(content)} 字符")

    # 修复1: 将 /* expr */("...") 替换为正确的输出语句
    def replace_expr_comments(match):
        expr_content = match.group(1)
        print(f"发现表达式注释: /* expr */{expr_content}")

        # 如果是字符串字面量，转换为输出语句
        if expr_content.startswith('("') and expr_content.endswith('")'):
            string_content = expr_content[1:-1]  # 去掉引号
            print(f"  -> 转换为输出语句: std::cout << \"{string_content}\" << std::endl")
            return f'std::cout << "{string_content}" << std::endl;'

        # 如果是变量或复杂表达式，使用通用处理
        print(f"  -> 转换为通用输出: std::cout << {expr_content} << std::endl")
        return f'std::cout << {expr_content} << std::endl;'

    # 应用修复
    pattern = r'/\* expr \*/\((.*?)\)'
    fixed_content = re.sub(pattern, replace_expr_comments, content)

    # 修复2: 添加缺失的include语句
    if '#include <iostream>' not in fixed_content:
        fixed_content = fixed_content.replace('#include <string>', '#include <string>\n#include <iostream>')
        print("添加缺失的 #include <iostream>")

    # 修复3: 移除main函数外的无效void __temp;声明
    def remove_invalid_temp_declarations(content):
        lines = content.split('\n')
        cleaned_lines = []
        in_main_function = False
        main_brace_count = 0

        for line in lines:
            stripped = line.strip()

            # 检测是否进入main函数
            if 'int main()' in stripped:
                in_main_function = True
                cleaned_lines.append(line)
                continue

            # 在main函数内，计数大括号
            if in_main_function:
                if '{' in stripped:
                    main_brace_count += 1
                if '}' in stripped:
                    main_brace_count -= 1
                    if main_brace_count == 0:
                        in_main_function = False

                # 只保留非void __temp;的行
                if stripped != 'void __temp;':
                    cleaned_lines.append(line)
            else:
                # 在main函数外，跳过void __temp;声明
                if stripped != 'void __temp;':
                    cleaned_lines.append(line)

        return '\n'.join(cleaned_lines)

    fixed_content = remove_invalid_temp_declarations(fixed_content)
    print("移除了无效的 void __temp; 声明")

    # 修复4: 确保main函数有正确的return语句
    if 'int main()' in fixed_content and 'return 0;' not in fixed_content:
        fixed_content = fixed_content.replace('}', '\n    return 0;\n}')
        print("添加缺失的 return 0;")

    # 统计修复内容
    changes_made = fixed_content != original_content
    print(f"是否进行了修改: {'是' if changes_made else '否'}")

    if changes_made:
        # 创建备份
        backup_path = cpp_file_path + '.backup'
        with open(backup_path, 'w') as f:
            f.write(original_content)
        print(f"原始文件已备份到: {backup_path}")

        # 写入修复后的内容
        with open(cpp_file_path, 'w') as f:
            f.write(fixed_content)

        print("=== C++代码修复完成 ===")
        print(f"修复后的文件: {cpp_file_path}")
        print(f"修复后文件大小: {len(fixed_content)} 字符")
        print("\n=== 修复后的代码 ===")
        print(fixed_content)
        print("======================")

        return True
    else:
        print("未发现需要修复的问题")
        return False

def compile_and_test(cpp_file_path):
    """编译并测试C++代码"""

    # 编译C++代码
    exe_path = cpp_file_path.replace('.cpp', '_compiled')
    compile_cmd = f'g++ -o {exe_path} {cpp_file_path}'

    print(f"\n=== 编译C++代码 ===")
    print(f"编译命令: {compile_cmd}")

    result = os.system(compile_cmd)
    if result != 0:
        print(f"编译失败，退出码: {result}")
        return False

    print("编译成功！")

    # 运行测试
    print(f"\n=== 运行C++程序 ===")
    print(f"运行命令: {exe_path}")

    result = os.system(exe_path)
    if result != 0:
        print(f"运行失败，退出码: {result}")
        return False

    print("运行成功！")
    return True

def main():
    """主函数"""

    if len(sys.argv) != 2:
        print("用法: python3 fix_cpp_code.py <cpp文件路径>")
        sys.exit(1)

    cpp_file = sys.argv[1]

    # 修复代码
    if fix_cpp_code(cpp_file):
        # 编译和测试
        compile_and_test(cpp_file)
    else:
        print("无需修复，直接编译和测试")
        compile_and_test(cpp_file)

if __name__ == "__main__":
    main()

File: fix_cpp_code.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


std::string fix_cpp_code(std::string cpp_file_path) {
    /* docstring */
    if (not os.path.exists(cpp_file_path)) {
        std::cout << to_str("错误：文件 ") << to_str(cpp_file_path) << to_str(" 不存在") << std::endl;
        return false;
    }
    {
        auto content = f.read();
    }
    auto original_content = content;
    std::cout << "=== 开始修复C++代码 ===" << std::endl;
    std::cout << to_str("原始文件: ") << to_str(cpp_file_path) << std::endl;
    std::cout << to_str("原始文件大小: ") << to_str(len(content)) << to_str(" 字符") << std::endl;
    std::function<int()> replace_expr_comments = [&]() {
        auto expr_content = match.group(1);
        std::cout << to_str("发现表达式注释: /* expr */") << to_str(expr_content) << std::endl;
        if (expr_content.startswith('("') and expr_content.endswith('")')) {
            auto string_content = expr_content[1:-1];
            std::cout << to_str("  -> 转换为输出语句: std::cout << \\\"") << to_str(string_content) << to_str("\\\" << std::endl") << std::endl;
            return to_str("std::cout << \"") + to_str(string_content) + to_str("\" << std::endl;");
        }
        std::cout << to_str("  -> 转换为通用输出: std::cout << ") << to_str(expr_content) << to_str(" << std::endl") << std::endl;
        return to_str("std::cout << ") + to_str(expr_content) + to_str(" << std::endl;");
    };
    auto pattern = r'/\* expr \*/\((.*?)\)';
    auto fixed_content = re.sub(pattern, replace_expr_comments, content);
    if ('#include <iostream>' not in fixed_content) {
        auto fixed_content = fixed_content.replace('#include <string>', '#include <string>\n#include <iostream>');
        std::cout << "添加缺失的 #include <iostream>" << std::endl;
    }
    std::function<int()> remove_invalid_temp_declarations = [&]() {
        auto lines = content.split('\n');
        auto cleaned_lines = std::vector<int>{};
        auto in_main_function = false;
        auto main_brace_count = 0;
        for (auto line : lines) {
            auto stripped = line.strip();
            if ('int main()' in stripped) {
                auto in_main_function = true;
                cleaned_lines.push_back(line);
                continue;
            }
            if (in_main_function) {
                if ('{' in stripped) {
                    main_brace_count += 1;
                }
                if ('}' in stripped) {
                    main_brace_count -= 1;
                    if (main_brace_count == 0) {
                        auto in_main_function = false;
                    }
                }
                if (stripped != 'void __temp;') {
                    cleaned_lines.push_back(line);
                }
            }
            else {
                if (stripped != 'void __temp;') {
                    cleaned_lines.push_back(line);
                }
            }
        }
        return '\n'.join(cleaned_lines);
    };
    auto fixed_content = remove_invalid_temp_declarations(fixed_content);
    std::cout << "移除了无效的 void __temp; 声明" << std::endl;
    if ('int main()' in fixed_content and 'return 0;' not in fixed_content) {
        auto fixed_content = fixed_content.replace('}', '\n    return 0;\n}');
        std::cout << "添加缺失的 return 0;" << std::endl;
    }
    auto changes_made = fixed_content != original_content;
    std::cout << to_str("是否进行了修改: ") << to_str((changes_made ? '是' : '否')) << std::endl;
    if (changes_made) {
        auto backup_path = cpp_file_path + '.backup';
        {
            f.write(original_content);
        }
        std::cout << to_str("原始文件已备份到: ") << to_str(backup_path) << std::endl;
        {
            f.write(fixed_content);
        }
        std::cout << "=== C++代码修复完成 ===" << std::endl;
        std::cout << to_str("修复后的文件: ") << to_str(cpp_file_path) << std::endl;
        std::cout << to_str("修复后文件大小: ") << to_str(len(fixed_content)) << to_str(" 字符") << std::endl;
        std::cout << "\n=== 修复后的代码 ===" << std::endl;
        std::cout << fixed_content << std::endl;
        std::cout << "======================" << std::endl;
        return true;
    }
    else {
        std::cout << "未发现需要修复的问题" << std::endl;
        return false;
    }
}


bool compile_and_test(std::string cpp_file_path) {
    /* docstring */
    auto exe_path = cpp_file_path.replace('.cpp', '_compiled');
    auto compile_cmd = to_str("g++ -o ") + to_str(exe_path) + to_str(" ") + to_str(cpp_file_path);
    std::cout << to_str("\\n=== 编译C++代码 ===") << std::endl;
    std::cout << to_str("编译命令: ") << to_str(compile_cmd) << std::endl;
    auto result = os.system(compile_cmd);
    if (result != 0) {
        std::cout << to_str("编译失败，退出码: ") << to_str(result) << std::endl;
        return false;
    }
    std::cout << "编译成功！" << std::endl;
    std::cout << to_str("\\n=== 运行C++程序 ===") << std::endl;
    std::cout << to_str("运行命令: ") << to_str(exe_path) << std::endl;
    auto result = os.system(exe_path);
    if (result != 0) {
        std::cout << to_str("运行失败，退出码: ") << to_str(result) << std::endl;
        return false;
    }
    std::cout << "运行成功！" << std::endl;
    return true;
}


int main() {
    /* docstring */
    if (len(sys::argv) != 2) {
        std::cout << "用法: python3 fix_cpp_code.py <cpp文件路径>" << std::endl;
        /* return (1; */
    }
    auto cpp_file = sys::argv[1];
    if (fix_cpp_code(cpp_file)) {
        compile_and_test(cpp_file);
    }
    else {
        std::cout << "无需修复，直接编译和测试" << std::endl;
        compile_and_test(cpp_file);
    }
    return 0;
}


